akka-persistence/src/test/scala/akka/persistence/SnapshotRecoveryWithEmptyJournalSpec.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence
+
+import akka.actor._
+import akka.testkit._
+
+object SnapshotRecoveryWithEmptyJournalSpec {
+  case class TakeSnapshot(deleteMessages: Boolean)
+
+  class SaveSnapshotTestPersistentActor(name: String, probe: ActorRef) extends NamedPersistentActor(name) {
+    var deleteMessages = false
+    var state = List.empty[String]
+
+    override def receiveRecover: Receive = {
+      case payload: String                     => state = s"${payload}-${lastSequenceNr}" :: state
+      case SnapshotOffer(_, snapshot: List[_]) => state = snapshot.asInstanceOf[List[String]]
+    }
+
+    override def receiveCommand = {
+      case payload: String =>
+        persist(payload) { _ =>
+          state = s"${payload}-${lastSequenceNr}" :: state
+        }
+      case TakeSnapshot(deleteMsgs) =>
+        deleteMessages = deleteMsgs
+        saveSnapshot(state)
+      case SaveSnapshotSuccess(md) =>
+        if (deleteMessages) {
+          deleteMessages(md.sequenceNr)
+          deleteMessages = false
+        }
+        probe ! md.sequenceNr
+      case GetState                 => probe ! state.reverse
+      case o: DeleteMessagesSuccess => probe ! o
+    }
+  }
+
+  class LoadSnapshotTestPersistentActor(name: String, _recovery: Recovery, probe: ActorRef)
+    extends NamedPersistentActor(name) {
+    override def recovery: Recovery = _recovery
+
+    override def receiveRecover: Receive = {
+      case payload: String      => probe ! s"${payload}-${lastSequenceNr}"
+      case offer: SnapshotOffer => probe ! offer
+      case other                => probe ! other
+    }
+
+    override def receiveCommand = {
+      case "done" => probe ! "done"
+      case payload: String =>
+        persist(payload) { _ =>
+          probe ! s"${payload}-${lastSequenceNr}"
+        }
+      case offer: SnapshotOffer => probe ! offer
+      case other                => probe ! other
+    }
+  }
+}
+
+class SnapshotRecoveryWithEmptyJournalSpec extends PersistenceSpec(PersistenceSpec.config(
+  "inmem", // with leveldb the error cannot be reproduced (counterKey is not reset after deleteMessages() )
```

Comment:
The change shouldn't be needed, because according to the journal protocol `asyncReadHighestSequenceNr` should not forgot about the highest used sequence number. This is a bug in the `InmemJournal`.

Where did you notice this problem originally? Which journal did you use?

Reply:
I see that you already said LeveldbJournal. That is correct behavior. The problem is in `InmemJournal`.

Reply:
Hi there, the test was a means to replicate the behavior that I have encountered when a LevelDBJournal is completely gone (see above for background info). I guess the proper 1:1 test would be to shutdown the ActorSystem, delete the Journal from disk and start up again. I'm happy to rewrite this if the proposed fix actually makes sense.

Reply:
ah, sorry for not reading your description fully

ok, it doesn't hurt use the highest from snapshot